### PR TITLE
Fix: NoT matching

### PR DESCRIPTION
**Pull request by EnnoMeijers**
Trap errormsg from NoT and improve matching.

### CODE REVIEW

File: termennetwerk.ts [hunk 1]
```diff
@@ -51,22 +52,33 @@ export async function getUriOfSearchTerm(sources: string[], searchTerm: string):
...
     const response: GraphQlResponse = await client.request(graphQlQuery, { sources, query: searchTerm });
-    if (!Array.isArray(response.terms) ||
-        response.terms[0].result.terms.length !== 1 ||
-        !response.terms[0].result ||
-        !response.terms[0].result.terms ||
-        !Array.isArray(response.terms[0].result.terms)) {
+    if (!Array.isArray(response.terms[0].result.terms) ||   // terms[] does not exists when an error occurs like a timeout
+        response.terms[0].result.terms.length !== 1 ) {     // allow only a single result
+          if (response.terms[0].result.message) {
+            console.error('Error message from Network-of-Terms API: ',response.terms[0].result.message,' for keyword: ', searchTerm);
```
Comment (sdevalk) [inline termennetwerk.ts]: Alternatively, you could use string Interpolation:

```
console.error(`Error message from Network-of-Terms API: "${result.message}" for keyword: "${searchTerm}"`);
```

File: termennetwerk.ts [hunk 2]
```diff
@@ -51,22 +52,33 @@ export async function getUriOfSearchTerm(sources: string[], searchTerm: string):
     const cacheKey = searchTerm.toLowerCase();
     if (cache[cacheKey]) return cache[cacheKey];
     const response: GraphQlResponse = await client.request(graphQlQuery, { sources, query: searchTerm });
-    if (!Array.isArray(response.terms) ||
-        response.terms[0].result.terms.length !== 1 ||
-        !response.terms[0].result ||
-        !response.terms[0].result.terms ||
-        !Array.isArray(response.terms[0].result.terms)) {
+    if (!Array.isArray(response.terms[0].result.terms) ||   // terms[] does not exists when an error occurs like a timeout
```
Comment (sdevalk) [inline termennetwerk.ts]: To make the code a bit easier on the eyes, you could first create a separate variable that holds the result:

`const result = response.terms[0].result;`

File: termennetwerk.ts [hunk 3]
```diff
@@ -51,22 +52,33 @@ export async function getUriOfSearchTerm(sources: string[], searchTerm: string):
...
+          if (response.terms[0].result.message) {
+            console.error('Error message from Network-of-Terms API: ',response.terms[0].result.message,' for keyword: ', searchTerm);
+          }
           return undefined;
     }
-
     /*
       searchTerm 'fietsen (transportmiddelen)' => http://vocab.getty.edu/aat/300212636 (prefLabel)
       searchTerm 'rijwiel' => http://vocab.getty.edu/aat/300212636 (altLabel)
       searchTerm 'rijwielen' => http://vocab.getty.edu/aat/300212636 (altLabel)
```
Comment (sdevalk) [inline termennetwerk.ts]: These three examples presumably worked before, but now they yield an `undefined`: there are 2 or more AAT terms that match these search words. Perhaps add some new, working examples?

File: termennetwerk.ts [hunk 4]
```diff
@@ -51,22 +52,33 @@ export async function getUriOfSearchTerm(sources: string[], searchTerm: string):
...
+      (label) => {
+         return label.toLowerCase() === searchTerm.toLowerCase();
+      } 
+    );
+
+    const inAltLabel = term.altLabel.find(
+      (label) => {
+        return label.toLowerCase() === searchTerm.toLowerCase();
+      } 
+    );
```
Comment (sdevalk) [inline termennetwerk.ts]: You could shorten this a bit with the code underneath:

```
const containsSearchTerm = (label: string) => label.toLowerCase() === searchTerm.toLowerCase();
const inPrefLabel = term.prefLabel.some(containsSearchTerm);
const inAltLabel = term.altLabel.some(containsSearchTerm);
```

**Comment by sdevalk** [comment]
@EnnoMeijers Perhaps we could also add the new `queryMode` input parameter to the GraphQL query, in order to trigger the new query pre-processor? Snippet:

```
query Terms($sources: [ID]! $query: String!) {
    terms(
      sources: $sources
      query: $query
      queryMode: OPTIMIZED # <-- this
    ) {
```

**Comment by EnnoMeijers** [comment]
Great suggestions Sjors, I will update the PR!

File: termennetwerk.ts (hunk 3 above)
Comment (EnnoMeijers) [reply termennetwerk.ts]: I removed the examples for now, maybe add a csv with examples for testing sometime? ;)

File: termennetwerk.ts [hunk 5]
```diff
@@ -51,22 +53,20 @@ export async function getUriOfSearchTerm(sources: string[], searchTerm: string):
     const cacheKey = searchTerm.toLowerCase();
     if (cache[cacheKey]) return cache[cacheKey];
     const response: GraphQlResponse = await client.request(graphQlQuery, { sources, query: searchTerm });
-    if (!Array.isArray(response.terms) ||
-        response.terms[0].result.terms.length !== 1 ||
-        !response.terms[0].result ||
-        !response.terms[0].result.terms ||
-        !Array.isArray(response.terms[0].result.terms)) {
+    const result = response.terms[0].result;
+    if (!Array.isArray(result.terms) ||   // terms[] does not exists when an error occurs like a timeout
```
Comment (ddeboer) [inline termennetwerk.ts:57]: Discriminating on the result type (`result.__typename === 'Terms'`) is cleaner and less prone to break than checking the `terms` property.

Comment (EnnoMeijers) [reply termennetwerk.ts:57]: I agree and I tried to that but it failed with:
` Property '__typename' does not exist on type '{ terms: { uri: string; prefLabel: string[]; altLabel: string[]; }[]; message: string; }'.` 
So it unclear to me how to do this properly.

Comment (ddeboer) [reply termennetwerk.ts:57]: That’s TypeScript saying that `__typename` is missing from the `GraphQlResponse` interface. It is currently only defined in `graphQlQuery`.

Comment (EnnoMeijers) [reply termennetwerk.ts:57]: Ah thanks, that helps!

**Pull request merged**